// @tanstack/react-router@1.34.3/dist/esm/useBlocker.js
import * as React from "react";
import { useRouter } from "./useRouter.js";
function useBlocker(blockerFnOrOpts, condition) {
  const { blockerFn, blockerCondition } = blockerFnOrOpts ? typeof blockerFnOrOpts === "function" ? { blockerFn: blockerFnOrOpts, blockerCondition: condition ?? true } : {
    blockerFn: blockerFnOrOpts.blockerFn,
    blockerCondition: blockerFnOrOpts.condition ?? true
  } : { blockerFn: void 0, blockerCondition: condition ?? true };
  const { history } = useRouter();
  const [resolver, setResolver] = React.useState({
    status: "idle",
    proceed: () => {
    },
    reset: () => {
    }
  });
  const createPromise = () => new Promise((resolve) => {
    setResolver({
      status: "idle",
      proceed: () => resolve(true),
      reset: () => resolve(false)
    });
  });
  const [promise, setPromise] = React.useState(createPromise);
  React.useEffect(() => {
    const blockerFnComposed = async () => {
      if (blockerFn) {
        return await blockerFn();
      }
      setResolver((prev) => ({
        ...prev,
        status: "blocked"
      }));
      const canNavigateAsync = await promise;
      setPromise(createPromise);
      return canNavigateAsync;
    };
    return !blockerCondition ? void 0 : history.block(blockerFnComposed);
  }, [blockerFn, blockerCondition, history, promise]);
  return resolver;
}
function Block({ blockerFn, condition, children }) {
  const resolver = useBlocker({ blockerFn, condition });
  return children ? typeof children === "function" ? children(resolver) : children : null;
}
export {
  Block,
  useBlocker
};
