// @tanstack/router-vite-plugin@1.34.1/dist/cjs/eliminateUnreferencedIdentifiers.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const t = require("@babel/types");
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const t__namespace = /* @__PURE__ */ _interopNamespaceDefault(t);
const eliminateUnreferencedIdentifiers = (programPath, refs) => {
  let referencesRemovedInThisPass;
  const shouldBeRemoved = (ident) => {
    if (isIdentifierReferenced(ident))
      return false;
    return true;
  };
  const sweepFunction = (path) => {
    const identifier = getIdentifier(path);
    if ((identifier == null ? void 0 : identifier.node) && shouldBeRemoved(identifier)) {
      ++referencesRemovedInThisPass;
      if (t__namespace.isAssignmentExpression(path.parentPath.node) || t__namespace.isVariableDeclarator(path.parentPath.node)) {
        path.parentPath.remove();
      } else {
        path.remove();
      }
    }
  };
  const sweepImport = (path) => {
    const local = path.get("local");
    if (shouldBeRemoved(local)) {
      ++referencesRemovedInThisPass;
      path.remove();
      if (path.parent.specifiers.length === 0) {
        path.parentPath.remove();
      }
    }
  };
  const handleObjectPattern = (pattern) => {
    const properties = pattern.get("properties");
    properties.forEach((property) => {
      if (property.node.type === "ObjectProperty") {
        const value = property.get("value");
        if (t__namespace.isIdentifier(value)) {
          if (shouldBeRemoved(value)) {
            property.remove();
          }
        } else if (t__namespace.isObjectPattern(value)) {
          handleObjectPattern(value);
        }
      } else if (t__namespace.isRestElement(property.node)) {
        const argument = property.get("argument");
        if (t__namespace.isIdentifier(argument) && shouldBeRemoved(argument)) {
          property.remove();
        }
      }
    });
  };
  do {
    referencesRemovedInThisPass = 0;
    programPath.scope.crawl();
    programPath.traverse({
      VariableDeclarator(path) {
        if (path.node.id.type === "Identifier") {
          const local = path.get("id");
          if (shouldBeRemoved(local)) {
            ++referencesRemovedInThisPass;
            path.remove();
          }
        } else if (path.node.id.type === "ObjectPattern") {
          handleObjectPattern(
            path.get("id")
          );
        } else if (path.node.id.type === "ArrayPattern") {
          const pattern = path.get("id");
          let hasRemoved = false;
          pattern.get("elements").forEach((element, index) => {
            let identifierPath;
            if (t__namespace.isIdentifier(element.node)) {
              identifierPath = element;
            } else if (t__namespace.isRestElement(element.node)) {
              identifierPath = element.get(
                "argument"
              );
            } else {
              return;
            }
            if (shouldBeRemoved(identifierPath)) {
              hasRemoved = true;
              pattern.node.elements[index] = null;
            }
          });
          if (hasRemoved && pattern.node.elements.every((element) => element === null)) {
            path.remove();
            ++referencesRemovedInThisPass;
          }
        }
      },
      FunctionDeclaration: sweepFunction,
      FunctionExpression: sweepFunction,
      ArrowFunctionExpression: sweepFunction,
      ImportSpecifier: sweepImport,
      ImportDefaultSpecifier: sweepImport,
      ImportNamespaceSpecifier: sweepImport
    });
  } while (referencesRemovedInThisPass);
};
function getIdentifier(path) {
  const parentPath = path.parentPath;
  if (parentPath.type === "VariableDeclarator") {
    const variablePath = parentPath;
    const name = variablePath.get("id");
    return name.node.type === "Identifier" ? name : null;
  }
  if (parentPath.type === "AssignmentExpression") {
    const variablePath = parentPath;
    const name = variablePath.get("left");
    return name.node.type === "Identifier" ? name : null;
  }
  if (path.node.type === "ArrowFunctionExpression") {
    return null;
  }
  if (path.node.type === "FunctionExpression") {
    return null;
  }
  return path.node.id && path.node.id.type === "Identifier" ? path.get("id") : null;
}
function isIdentifierReferenced(ident) {
  const binding = ident.scope.getBinding(ident.node.name);
  if (binding == null ? void 0 : binding.referenced) {
    if (binding.path.type === "FunctionDeclaration") {
      return !binding.constantViolations.concat(binding.referencePaths).every((ref) => ref.findParent((parent) => parent === binding.path));
    }
    return true;
  }
  return false;
}
exports.eliminateUnreferencedIdentifiers = eliminateUnreferencedIdentifiers;
